ompi/dpm/dpm.c

```diff
@@ -651,7 +651,7 @@ static int dpm_convert(opal_list_t *infos,
                        bool deprecated)
 {
     opal_info_item_t *iptr;
-    char *ck, *ptr, *help_str;
+    char *ck, *ptr, *help_str = NULL;
```

Comment:
This is why proper variable scoping makes code much more readable: if it was properly scoped [here](https://github.com/open-mpi/ompi/pull/8203/files#diff-a4f4102ea0e8f98faed7986c7a410f048dd1305fbf082046ef893c8aa4e3d3a6R758) we would not need the initialization here. And my alarm bells wouldn't have gone off seeing that there was another free [here](https://github.com/open-mpi/ompi/pull/8203/files#diff-a4f4102ea0e8f98faed7986c7a410f048dd1305fbf082046ef893c8aa4e3d3a6R693) without setting it to `NULL` (of course, it's not critical because there is a return as well...)

Reply:
FWIW, the compilers are doing more and more static analysis these days.  This particular fix came from the following warning from MacOS clang:

```
2361 dpm/dpm.c:757:8: warning: variable 'help_str' is used uninitialized whenever 'if' condition is false [-Wsometimes-uninitialized]
2362     if(deprecated) {
2363        ^~~~~~~~~~
2364 dpm/dpm.c:762:10: note: uninitialized use occurs here
2365     free(help_str);
2366          ^~~~~~~~
2367 ../opal/include/opal_config_bottom.h:352:34: note: expanded from macro 'free'
2368 #    define free(ptr) opal_free((ptr), __FILE__, __LINE__)
2369                                  ^~~
2370 dpm/dpm.c:757:5: note: remove the 'if' if its condition is always true
2371     if(deprecated) {
2372     ^~~~~~~~~~~~~~~
2373 dpm/dpm.c:654:30: note: initialize the variable 'help_str' to silence this warning
2374     char *ck, *ptr, *help_str;
2375                              ^
2376                               = NULL
```